File: agenttel-node/src/baseline/rolling.ts
import type { OperationBaseline, RollingSnapshot } from '../interfaces.js';
import type { BaselineProvider } from './provider.js';

/** A ring buffer for latency and error tracking. */
export class RollingWindow {
  private readonly values: Float64Array;
  private readonly times: Float64Array;
  private readonly errors: Uint8Array;
  private readonly capacity: number;
  private writeIdx = 0;
  private count = 0;

  constructor(capacity = 1000) {
    this.capacity = Math.max(1, capacity);
    this.values = new Float64Array(this.capacity);
    this.times = new Float64Array(this.capacity);
    this.errors = new Uint8Array(this.capacity);
  }

  record(value: number): void {
    const idx = this.writeIdx % this.capacity;
    this.values[idx] = value;
    this.times[idx] = Date.now();
    this.errors[idx] = 0;
    this.writeIdx++;
    this.count++;
  }

  recordError(): void {
    const idx = this.writeIdx % this.capacity;
    this.values[idx] = 0;
    this.times[idx] = Date.now();
    this.errors[idx] = 1;
    this.writeIdx++;
    this.count++;
  }

  size(): number {
    return Math.min(this.count, this.capacity);
  }

  snapshot(): RollingSnapshot {
    const n = this.size();
    if (n === 0) {
      return { mean: 0, stddev: 0, p50: 0, p95: 0, p99: 0, errorRate: 0, sampleCount: 0, ageMs: 0 };
    }

    const latencies: number[] = [];
    let errorCount = 0;
    let oldestTime = Infinity;

    for (let i = 0; i < n; i++) {
      if (this.errors[i]) {
        errorCount++;
      } else if (this.values[i] > 0 || this.times[i] > 0) {
        latencies.push(this.values[i]);
      }
      if (this.times[i] > 0 && this.times[i] < oldestTime) {
        oldestTime = this.times[i];
      }
    }

    if (latencies.length === 0) {
      return {
        mean: 0, stddev: 0, p50: 0, p95: 0, p99: 0,
        errorRate: 1.0,
        sampleCount: n,
        ageMs: oldestTime === Infinity ? 0 : Date.now() - oldestTime,
      };
    }

    latencies.sort((a, b) => a - b);
    const m = mean(latencies);
    const sd = stddev(latencies, m);

    return {
      mean: m,
      stddev: sd,
      p50: percentile(latencies, 0.50),
      p95: percentile(latencies, 0.95),
      p99: percentile(latencies, 0.99),
      errorRate: errorCount / n,
      sampleCount: n,
      ageMs: oldestTime === Infinity ? 0 : Date.now() - oldestTime,
    };
  }
}

/** Tracks rolling baselines per operation. */
export class RollingBaselineProvider implements BaselineProvider {
  private readonly windows = new Map<string, RollingWindow>();
  private readonly windowSize: number;
  private readonly minSamples: number;

  constructor(windowSize = 1000, minSamples = 10) {
    this.windowSize = Math.max(1, windowSize);
    this.minSamples = Math.max(1, minSamples);
  }

  recordLatency(operationName: string, latencyMs: number): void {
    this.getOrCreate(operationName).record(latencyMs);
  }

  recordError(operationName: string): void {
    this.getOrCreate(operationName).recordError();
  }

  getBaseline(operationName: string): OperationBaseline | undefined {
    const w = this.windows.get(operationName);
    if (!w) return undefined;

    const snap = w.snapshot();
    if (snap.sampleCount < this.minSamples) return undefined;

    return {
      operationName,
      latencyP50Ms: snap.p50,
      latencyP99Ms: snap.p99,
      errorRate: snap.errorRate,
      source: 'rolling_7d',
      updatedAt: new Date(),
    };
  }

  getSnapshot(operationName: string): RollingSnapshot | undefined {
    const w = this.windows.get(operationName);
    if (!w) return undefined;
    return w.snapshot();
  }

  private getOrCreate(operationName: string): RollingWindow {
    let w = this.windows.get(operationName);
    if (!w) {
      w = new RollingWindow(this.windowSize);
      this.windows.set(operationName, w);
    }
    return w;
  }
}

function mean(vals: number[]): number {
  if (vals.length === 0) return 0;
  let sum = 0;
  for (const v of vals) sum += v;
  return sum / vals.length;
}

function stddev(vals: number[], m: number): number {
  if (vals.length < 2) return 0;
  let sumSq = 0;
  for (const v of vals) {
    const d = v - m;
    sumSq += d * d;
  }
  return Math.sqrt(sumSq / vals.length);
}

function percentile(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0;
  const idx = p * (sorted.length - 1);
  const lower = Math.floor(idx);
  const upper = Math.ceil(idx);
  if (lower === upper || upper >= sorted.length) return sorted[lower];
  const frac = idx - lower;
  return sorted[lower] * (1 - frac) + sorted[upper] * frac;
}

File: agenttel-go/baseline/rolling.go
package baseline

import (
	"math"
	"sort"
	"sync"
	"sync/atomic"
	"time"

	"go.agenttel.dev/agenttel/models"
)

// RollingWindow is a lock-free ring buffer for latency and error tracking.
type RollingWindow struct {
	values    []float64
	times     []int64 // unix nano
	errors    []bool
	capacity  int
	writeIdx  atomic.Int64
	count     atomic.Int64
	mu        sync.RWMutex // for snapshot reads
}

// NewRollingWindow creates a ring buffer with the given capacity.
func NewRollingWindow(capacity int) *RollingWindow {
	if capacity <= 0 {
		capacity = 1000
	}
	return &RollingWindow{
		values:   make([]float64, capacity),
		times:    make([]int64, capacity),
		errors:   make([]bool, capacity),
		capacity: capacity,
	}
}

// Record adds a latency sample to the window.
func (w *RollingWindow) Record(value float64) {
	idx := int(w.writeIdx.Add(1)-1) % w.capacity
	w.mu.Lock()
	w.values[idx] = value
	w.times[idx] = time.Now().UnixNano()
	w.errors[idx] = false
	w.mu.Unlock()
	w.count.Add(1)
}

// RecordError adds an error sample.
func (w *RollingWindow) RecordError() {
	idx := int(w.writeIdx.Add(1)-1) % w.capacity
	w.mu.Lock()
	w.values[idx] = 0
	w.times[idx] = time.Now().UnixNano()
	w.errors[idx] = true
	w.mu.Unlock()
	w.count.Add(1)
}

// Size returns the number of samples recorded (capped at capacity).
func (w *RollingWindow) Size() int {
	n := int(w.count.Load())
	if n > w.capacity {
		return w.capacity
	}
	return n
}

// Snapshot returns a point-in-time statistical snapshot of the window.
func (w *RollingWindow) Snapshot() models.RollingSnapshot {
	w.mu.RLock()
	defer w.mu.RUnlock()

	n := w.Size()
	if n == 0 {
		return models.RollingSnapshot{}
	}

	// Collect non-error latency values
	latencies := make([]float64, 0, n)
	errorCount := 0
	var oldestTime int64 = math.MaxInt64

	for i := 0; i < n; i++ {
		if w.errors[i] {
			errorCount++
		} else if w.values[i] > 0 || w.times[i] > 0 {
			latencies = append(latencies, w.values[i])
		}
		if w.times[i] > 0 && w.times[i] < oldestTime {
			oldestTime = w.times[i]
		}
	}

	if len(latencies) == 0 {
		return models.RollingSnapshot{
			ErrorRate:   1.0,
			SampleCount: n,
			AgeMs:       ageMs(oldestTime),
		}
	}

	sort.Float64s(latencies)
	mean := mean(latencies)
	stddev := stddev(latencies, mean)

	return models.RollingSnapshot{
		Mean:        mean,
		Stddev:      stddev,
		P50:         percentile(latencies, 0.50),
		P95:         percentile(latencies, 0.95),
		P99:         percentile(latencies, 0.99),
		ErrorRate:   float64(errorCount) / float64(n),
		SampleCount: n,
		AgeMs:       ageMs(oldestTime),
	}
}

func mean(vals []float64) float64 {
	if len(vals) == 0 {
		return 0
	}
	sum := 0.0
	for _, v := range vals {
		sum += v
	}
	return sum / float64(len(vals))
}

func stddev(vals []float64, mean float64) float64 {
	if len(vals) < 2 {
		return 0
	}
	sumSq := 0.0
	for _, v := range vals {
		d := v - mean
		sumSq += d * d
	}
	return math.Sqrt(sumSq / float64(len(vals)))
}

func percentile(sorted []float64, p float64) float64 {
	if len(sorted) == 0 {
		return 0
	}
	idx := p * float64(len(sorted)-1)
	lower := int(math.Floor(idx))
	upper := int(math.Ceil(idx))
	if lower == upper || upper >= len(sorted) {
		return sorted[lower]
	}
	frac := idx - float64(lower)
	return sorted[lower]*(1-frac) + sorted[upper]*frac
}

func ageMs(oldestNano int64) int64 {
	if oldestNano == math.MaxInt64 || oldestNano == 0 {
		return 0
	}
	return (time.Now().UnixNano() - oldestNano) / int64(time.Millisecond)
}

// RollingProvider tracks rolling baselines per operation.
type RollingProvider struct {
	windows    map[string]*RollingWindow
	mu         sync.RWMutex
	windowSize int
	minSamples int
}

// NewRollingProvider creates a RollingProvider.
func NewRollingProvider(windowSize, minSamples int) *RollingProvider {
	if windowSize <= 0 {
		windowSize = 1000
	}
	if minSamples <= 0 {
		minSamples = 10
	}
	return &RollingProvider{
		windows:    make(map[string]*RollingWindow),
		windowSize: windowSize,
		minSamples: minSamples,
	}
}

// RecordLatency adds a latency sample for an operation.
func (r *RollingProvider) RecordLatency(operationName string, latencyMs float64) {
	r.getOrCreate(operationName).Record(latencyMs)
}

// RecordError adds an error sample for an operation.
func (r *RollingProvider) RecordError(operationName string) {
	r.getOrCreate(operationName).RecordError()
}

// GetBaseline returns a computed baseline if sufficient samples exist.
func (r *RollingProvider) GetBaseline(operationName string) (models.OperationBaseline, bool) {
	r.mu.RLock()
	w, ok := r.windows[operationName]
	r.mu.RUnlock()

	if !ok {
		return models.OperationBaseline{}, false
	}

	snap := w.Snapshot()
	if snap.SampleCount < r.minSamples {
		return models.OperationBaseline{}, false
	}

	return models.OperationBaseline{
		OperationName: operationName,
		LatencyP50Ms:  snap.P50,
		LatencyP99Ms:  snap.P99,
		ErrorRate:     snap.ErrorRate,
		Source:        "rolling_7d",
		UpdatedAt:     time.Now(),
	}, true
}

// GetSnapshot returns the raw rolling snapshot for an operation.
func (r *RollingProvider) GetSnapshot(operationName string) (models.RollingSnapshot, bool) {
	r.mu.RLock()
	w, ok := r.windows[operationName]
	r.mu.RUnlock()

	if !ok {
		return models.RollingSnapshot{}, false
	}
	return w.Snapshot(), true
}

func (r *RollingProvider) getOrCreate(operationName string) *RollingWindow {
	r.mu.RLock()
	w, ok := r.windows[operationName]
	r.mu.RUnlock()

	if ok {
		return w
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	// Double-check after acquiring write lock
	if w, ok = r.windows[operationName]; ok {
		return w
	}

	w = NewRollingWindow(r.windowSize)
	r.windows[operationName] = w
	return w
}
